File: rxbinding-support-v4/src/main/java/com/jakewharton/rxbinding2/support/v4/widget/DrawerLayoutDrawerOpenedObservable.java
package com.jakewharton.rxbinding2.support.v4.widget;

import android.support.v4.widget.DrawerLayout;
import android.support.v4.widget.DrawerLayout.DrawerListener;
import android.view.View;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static io.reactivex.android.MainThreadDisposable.verifyMainThread;

final class DrawerLayoutDrawerOpenedObservable extends Observable<Boolean> {
  private final DrawerLayout view;
  private final int gravity;

  DrawerLayoutDrawerOpenedObservable(DrawerLayout view, int gravity) {
    this.view = view;
    this.gravity = gravity;
  }

  @Override protected void subscribeActual(Observer<? super Boolean> observer) {
    verifyMainThread();
    Listener listener = new Listener(view, gravity, observer);
    observer.onSubscribe(listener);
    view.addDrawerListener(listener);
    observer.onNext(view.isDrawerOpen(gravity));
  }

  static final class Listener extends MainThreadDisposable implements DrawerListener {
    private final DrawerLayout view;
    private final int gravity;
    private final Observer<? super Boolean> observer;

    Listener(DrawerLayout view, int gravity, Observer<? super Boolean> observer) {
      this.view = view;
      this.gravity = gravity;
      this.observer = observer;
    }

    @Override public void onDrawerSlide(View drawerView, float slideOffset) {

    }

    @Override public void onDrawerOpened(View drawerView) {
      if (!isDisposed()) {
        int drawerGravity = ((DrawerLayout.LayoutParams) drawerView.getLayoutParams()).gravity;
        if (drawerGravity == gravity) {
          observer.onNext(true);
        }
      }
    }

    @Override public void onDrawerClosed(View drawerView) {
      if (!isDisposed()) {
        int drawerGravity = ((DrawerLayout.LayoutParams) drawerView.getLayoutParams()).gravity;
        if (drawerGravity == gravity) {
          observer.onNext(false);
        }
      }
    }

    @Override public void onDrawerStateChanged(int newState) {

    }

    @Override protected void onDispose() {
      view.removeDrawerListener(this);
    }
  }
}

File: rxbinding-drawerlayout/src/main/java/com/jakewharton/rxbinding4/drawerlayout/DrawerLayoutDrawerOpenedObservable.kt
@file:JvmName("RxDrawerLayout")
@file:JvmMultifileClass

package com.jakewharton.rxbinding4.drawerlayout

import android.view.View
import androidx.annotation.CheckResult
import androidx.drawerlayout.widget.DrawerLayout
import androidx.drawerlayout.widget.DrawerLayout.DrawerListener
import com.jakewharton.rxbinding4.InitialValueObservable
import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.android.MainThreadDisposable

import com.jakewharton.rxbinding4.internal.checkMainThread

/**
 * Create an observable of the open state of the drawer of `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 *
 * *Note:* A value will be emitted immediately on subscribe.
 */
@CheckResult
fun DrawerLayout.drawerOpen(gravity: Int): InitialValueObservable<Boolean> {
  return DrawerLayoutDrawerOpenedObservable(this, gravity)
}

private class DrawerLayoutDrawerOpenedObservable(
  private val view: DrawerLayout,
  private val gravity: Int
) : InitialValueObservable<Boolean>() {

  override fun subscribeListener(observer: Observer<in Boolean>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, gravity, observer)
    observer.onSubscribe(listener)
    view.addDrawerListener(listener)
  }

  override val initialValue get() = view.isDrawerOpen(gravity)

  private class Listener(
    private val view: DrawerLayout,
    private val gravity: Int,
    private val observer: Observer<in Boolean>
  ) : MainThreadDisposable(), DrawerListener {

    override fun onDrawerSlide(drawerView: View, slideOffset: Float) {
    }

    override fun onDrawerOpened(drawerView: View) {
      if (!isDisposed) {
        val drawerGravity = (drawerView.layoutParams as DrawerLayout.LayoutParams).gravity
        if (drawerGravity == gravity) {
          observer.onNext(true)
        }
      }
    }

    override fun onDrawerClosed(drawerView: View) {
      if (!isDisposed) {
        val drawerGravity = (drawerView.layoutParams as DrawerLayout.LayoutParams).gravity
        if (drawerGravity == gravity) {
          observer.onNext(false)
        }
      }
    }

    override fun onDrawerStateChanged(newState: Int) {
    }

    override fun onDispose() {
      view.removeDrawerListener(this)
    }
  }
}
